src/asl.ts

```diff
@@ -2129,33 +2319,152 @@ export class ASL {
+                      leftOutput.condition,
+              };
+            }
+          }
+
+          const condition =
+            /**
+             * Left is a condition
+             * ??: checks for null or undefined, a condition will always be true or false
+             * &&: the right will be run when the result is true
+             * ||: the right will be run when the result is false
+             */
+            ASLGraph.isConditionOutput(leftOutput)
+              ? expr.op === "??"
+                ? ASL.trueCondition()
+                : expr.op === "&&"
+                ? leftOutput.condition
+                : ASL.not(leftOutput.condition)
+              : /**
+               * Left is a a json path
+               * ??: check for undefined or null
+               * &&: the right will be run when the result is truthy
+               * ||: the right will be run when the result is falsy
+               */
+              expr.op === "||"
+              ? ASL.isTruthy(leftOutput.jsonPath)
+              : expr.op === "&&"
+              ? ASL.not(ASL.isTruthy(leftOutput.jsonPath))
+              : ASL.isPresentAndNotNull(leftOutput.jsonPath);
+
+          return this.conditionState(expr, condition, leftOutput, right);
+        });
+      } else if (expr.op === "??=" || expr.op === "&&=" || expr.op === "||=") {
+        if (!isVariableReference(expr.left)) {
+          throw new SynthError(
+            ErrorCodes.Unexpected_Error,
+            "Expected left side of assignment to be a variable."
+          );
+        }
+        return this.evalExpr(expr.left, (left) => {
+          const right = this.eval(expr.right);
+
+          if (!ASLGraph.isJsonPath(left)) {
+            throw new SynthError(
+              ErrorCodes.Unexpected_Error,
+              `Expected assignment to target a variable, found: ${
+                ASLGraph.isLiteralValue(left) ? left.value : "boolean"
+              }`
+            );
+          }
+
+          /**
+           * &&: left ? right : left
+           * ||: left ? left : right
+           * ??: left !== undefined && left !== null ? left : right
+           */
+          const condition =
+            expr.op === "||="
+              ? ASL.isTruthy(left.jsonPath)
+              : expr.op === "&&="
+              ? ASL.not(ASL.isTruthy(left.jsonPath))
+              : ASL.isPresentAndNotNull(left.jsonPath);
+
+          return this.conditionState(
+            expr,
+            condition,
+            left,
+            right,
+            left.jsonPath
+          );
+        });
+      }
+      assertNever(expr.op);
+    } else if (isAwaitExpr(expr)) {
+      return this.eval(expr.expr);
+    } else if (isTypeOfExpr(expr)) {
+      return this.evalExpr(expr.expr, (exprOutput) => {
+        if (ASLGraph.isLiteralValue(exprOutput)) {
+          return {
+            value: typeof exprOutput.value,
+            containsJsonPath: false,
+          };
```

Comment:
aside: do we support `Array.isArray`?

Reply:
Nope, and neither does ASL... we can do isPresent on `$.arr[0]` (I assume this won't error), but a zero sized array would also fail. As far as I know, there are no properties on an array we could check for...

Reply:
Could we Stringify it as JSON and check that somehow? `isPresent(a[0]) || ToJson(a) === "[]"`

Reply:
that may work. won't be pretty because we'll need a state to stingily and another to check the output.

Base case: is array of size 1 - 1 state (choice with condition)
Worst case: is not array or is empty - (choice: check for [1], pass: string format, choice: check for "[]")

We can optimize by checking for all other types in the first choice which would mean only objects would hit the worst case.